### PR TITLE
[llvm][cas] Make it an error to --ingest the CAS itself

### PR DESCRIPTION
**Pull request by benlangmuir**
Ingesting the cas into itself is probably never intended, and it creates a strange result since the CAS is being mutated while it happens. Moreover, when CAS files are large it can double or more the size of the CAS.

### CODE REVIEW

**Comment by benlangmuir** [comment]
@swift-ci please test

File: clang/test/CAS/cas-backend.c [hunk 1]
```diff
@@ -1,8 +1,8 @@
-// RUN: rm -rf %t && mkdir -p %t
-// RUN: llvm-cas --cas %t/cas --ingest --data %s > %t/casid
+// RUN: rm -rf %t %t.cas && mkdir -p %t
+// RUN: llvm-cas --cas %t.cas --ingest --data %s > %t/casid
```
Comment (akyrtzi) [inline clang/test/CAS/cas-backend.c]: This test doesn't need change, it's ingesting `%s`, was changing it intentional for consistency? There are some other tests like this as well.

File: clang/test/CAS/fcache-compile-job-dependency-file.c [hunk 2]
```diff
@@ -1,9 +1,9 @@
-// RUN: rm -rf %t && mkdir -p %t
+// RUN: rm -rf %t %t.cas && mkdir -p %t
 // RUN: split-file %s %t/src
-// RUN: llvm-cas --cas %t/cas --ingest --data %t/src > %t/casid
+// RUN: llvm-cas --cas %t.cas --ingest --data %t/src > %t/casid
```
Comment (akyrtzi) [inline clang/test/CAS/fcache-compile-job-dependency-file.c]: This test also doesn't need change.

**Review by akyrtzi — APPROVED** [review]
A question about the test changes, otherwise LGTM.

File: clang/test/CAS/cas-backend.c (hunk 1 above)
Comment (benlangmuir) [reply clang/test/CAS/cas-backend.c]: Oh, I saw so many failures inclang/test/CAS that I assumed every test that used llvm-cas --ingest was failing, so I just changed them all. I can change these two back if you like, though it seems harmless.

Comment (akyrtzi) [reply clang/test/CAS/cas-backend.c]: To be precise, half of these tests do not need changing, but it's up to you, I don't particularly care either way.

Comment (benlangmuir) [reply clang/test/CAS/cas-backend.c]: Hmm, you're right. Not sure what I was seeing before. Reverted the unnecessary changes.

**Comment by benlangmuir** [comment]
@swift-ci please test

File: llvm/tools/llvm-cas/llvm-cas.cpp [hunk 3]
```diff
@@ -457,11 +461,31 @@ static Expected<ObjectProxy> ingestFileSystemImpl(ObjectStore &CAS,
...
-int ingestFileSystem(ObjectStore &CAS, StringRef Path) {
+/// Check that we are not attempting to ingest the CAS into itself, which can
+/// accidentally create a weird or large cas.
+Error checkCASIngestPath(StringRef CASPath, StringRef DataPath) {
+  SmallString<128> RealCAS, RealData;
+  if (std::error_code EC = sys::fs::real_path(StringRef(CASPath), RealCAS))
+    return createFileError(CASPath, EC);
+  if (std::error_code EC = sys::fs::real_path(StringRef(DataPath), RealData))
+    return createFileError(CASPath, EC);
+  if (RealCAS.startswith(RealData) &&
```
Comment (cachemeifyoucan) [inline llvm/tools/llvm-cas/llvm-cas.cpp:472]: There is a `llvm::sys::path::starts_with` but I guess that one also doesn't support things like case insensitive system or the separator check.

**Review by cachemeifyoucan — APPROVED** [review]

File: llvm/tools/llvm-cas/llvm-cas.cpp (hunk 3 above)
Comment (benlangmuir) [reply llvm/tools/llvm-cas/llvm-cas.cpp:472]: I don't see such a function in sys::path.  There is `static bool starts_with` in Path.cpp, but it is just doing `StringRef::starts_with` if it's a unix path.  This probably *should* be an API though.

**Pull request merged**